File: demos/golang_demos/transaction/htdf_transfer.go
/**
date: 2020/9/25
author : yqq
descriptions: rewrite  HTDF transaction demo in golang
reference : https://gitee.com/orientwalt/apidoc_2020/blob/master/demo/python/htdf_transfer.py
core signature algorithm : secp256k1   (same as bitcoin , ethereum )
*/

package main

import (
	"bytes"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"io/ioutil"
	"math"
	"net/http"
	"strconv"
	"strings"
	"time"

	secp256k1 "github.com/btcsuite/btcd/btcec"
)

// this is a testchain  address that has 100HTDF for test, you can run this demo directly , 2020-09-25
var gChainId = "testchain" // mainchain or testchain
var gFromAddress = "htdf1xwpsq6yqx0zy6grygy7s395e2646wggufqndml"
var gPrivateKey = "279bdcd8dccec91f9e079894da33d6888c0f9ef466c0b200921a1bf1ea7d86e8"
var gNodeIpPort = "htdf2020-test01.orientwalt.cn:1317"
var gGasLimit = 30000
var gGasPrice = 100

// mainchain parameters
//var gChainId = "mainchain"                                     //TODO:  mainchain or testchain
//var gFromAddress = "YOUR_HTDF_ADDRESS"                         //TODO
//var gPrivateKey = "YOUR_HTDF_ADDRESS_PRIVATE_KEY"              //TODO
//var gNodeIpPort = "YOUR_HTDF_FULL_NODE_IP_ADDRESS:1317" 		 //TODO
//var gGasLimit = 30000
//var gGasPrice =  100

var gUnsignedStr = `{
	"account_number": "%d",
	"chain_id": "%s",
	"fee": {
		"gas_price": "%d",
		"gas_wanted": "%d"
	},
	"memo": "%s",
	"msgs": [{
		"Amount": [{
		"amount": "%d",
		"denom": "satoshi"
	}],
		"Data": "",
		"From": "%s",
		"GasPrice": %d,
		"GasWanted": %d,
		"To": "%s"
	}],
	"sequence": "%d"
	}`

var gBroadcastStr = `{
	"type": "auth/StdTx",
		"value":{
		"msg": [{
			"type": "htdfservice/send",
			"value":{
				"From": "%s",
				"To": "%s",
				"Amount": [{
					"denom": "satoshi",
					"amount": "%d"
				}],
				"Data": "",
				"GasPrice": "%d",
				"GasWanted": "%d"
			}
		}],
	"fee": {
		"gas_wanted": "%d",
		"gas_price": "%d"
	},
	"signatures": [{
		"pub_key": {
		"type": "tendermint/PubKeySecp256k1",
		"value": "%s"
	},
	"signature": "%s"
	}],
	"memo": "%s"
	}
	}`

func HtdfToSatoshi(amount float64) int64 {
	// 1 HTDF  = 10^8 satoshi
	return int64(amount * math.Pow(10, 8))
}

func Sign(privateKey []byte, msg []byte) (string, error) {
	pk, _ := secp256k1.PrivKeyFromBytes(secp256k1.S256(), privateKey[:])

	sig, err := pk.Sign(msg)
	if err != nil {
		return "", err
	}
	sigRS := hex.EncodeToString(sig.R.Bytes()) + hex.EncodeToString(sig.S.Bytes())
	fmt.Printf("rs: %s\n", sigRS)
	return sigRS, nil
}

func HttpGet(url string) (string, error) {
	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Get(url)
	if err != nil {
		return "", err
	}
	if resp == nil {
		return "", fmt.Errorf("empty response ")
	}
	defer resp.Body.Close()

	if 200 != resp.StatusCode {
		// HTTP 204  if an account has no transaction
		if 204 == resp.StatusCode {
			return "", fmt.Errorf("not found any information")
		}
		return "", fmt.Errorf("http code: %v", resp.StatusCode)
	}

	var buffer [1024]byte
	result := bytes.NewBuffer(nil)
	for {
		n, err := resp.Body.Read(buffer[0:])
		result.Write(buffer[0:n])
		if err != nil && err == io.EOF {
			break
		} else if err != nil {
			return "", err
		}
	}

	return result.String(), nil
}

func HttpPost(url string, data interface{}, contentType string) (string, error) {
	client := &http.Client{Timeout: 10 * time.Second}
	jsonStr, _ := json.Marshal(data)
	resp, err := client.Post(url, contentType, bytes.NewBuffer(jsonStr))
	if err != nil {
		return "", err
	}
	if resp == nil {
		return "", fmt.Errorf("empty response ")
	}
	defer resp.Body.Close()

	result, err := ioutil.ReadAll(resp.Body)
	return string(result), err
}

func Transfer(amount int64, dstAddress string, memo string) error {

	fmt.Println("================ Step 0 : check parameters ")
	if amount < 1 || amount > (HtdfToSatoshi(60_000_000)) {
		return fmt.Errorf("invalid amount %v\n", amount)
	}

	// in this demo, `memo` DO NOT contains '\t' or '\r' or '\n'  or  space in memo
	if len(memo) > 256 || strings.ContainsAny(memo, "\t\r\n ") {
		return fmt.Errorf("memo is illegal")
	}

	//TODO: check dstAddress
	// reference: https://gitee.com/orientwalt/apidoc_2020/blob/master/demo/golang_bech32/bech32.go

	fmt.Println("==================== Step 1 : get account details by RPC")
	url := fmt.Sprintf("http://%s/auth/accounts/%s", gNodeIpPort, gFromAddress)
	data, err := HttpGet(url)
	if nil != err {
		return err
	}

	fmt.Printf("%s\n", data)
	var m map[string]interface{}
	if nil != json.Unmarshal([]byte(data), &m) {
		return err
	}
	mapValue := m["value"].(map[string]interface{})
	coins := mapValue["coins"].([]interface{})
	balance := coins[0].(map[string]interface{})["amount"].(string)
	accountNumber, err := strconv.Atoi(mapValue["account_number"].(string))
	if nil != err {
		return err
	}
	sequence, err := strconv.Atoi(mapValue["sequence"].(string))
	if nil != err {
		return err
	}
	fmt.Printf("account details:  accountNumber : %v, balance: %v, sequence: %v\n",
		accountNumber, balance, sequence)

	fmt.Println(" ===================== Step 2 : make raw transaction string")
	fmtUnsignedStr := fmt.Sprintf(gUnsignedStr, accountNumber, gChainId, gGasPrice,
		gGasLimit, memo, amount, gFromAddress, gGasPrice, gGasLimit, dstAddress, sequence)

	fmtUnsignedStr = strings.Replace(fmtUnsignedStr, " ", "", -1)
	fmtUnsignedStr = strings.Replace(fmtUnsignedStr, "\t", "", -1)
	fmtUnsignedStr = strings.Replace(fmtUnsignedStr, "\n", "", -1)
	fmtUnsignedStr = strings.Replace(fmtUnsignedStr, "\r", "", -1)

	fmt.Printf("formated raw transaction string: %v", fmtUnsignedStr)

	fmt.Printf("======================= Step 3 : make signature \n")
	hash := sha256.New()
	hash.Write([]byte(fmtUnsignedStr))
	msgHash := hash.Sum(nil)
	fmt.Printf("msgHash: %v\n", hex.EncodeToString(msgHash))

	pkBytes, err := hex.DecodeString(gPrivateKey)
	if nil != err {
		return err
	}

	sig, err := Sign(pkBytes, msgHash)
	if nil != err {
		return err
	}
	fmt.Printf("signature: %v\n", sig)

	fmt.Println("===================== Step 4 : make broadcast json string")
	sigBytes, err := hex.DecodeString(sig)
	if nil != err {
		return err
	}
	b64Sig := base64.StdEncoding.EncodeToString(sigBytes)
	fmt.Printf("base64 encoded signature : %v\n", b64Sig)

	_, pubKey := secp256k1.PrivKeyFromBytes(secp256k1.S256(), pkBytes[:])
	b64PubKey := base64.StdEncoding.EncodeToString(pubKey.SerializeCompressed())
	fmt.Printf("base64 encoded public key: %v", b64PubKey)

	fmtBroadcastStr := fmt.Sprintf(gBroadcastStr, gFromAddress, dstAddress, amount,
		gGasPrice, gGasLimit, gGasLimit, gGasPrice, b64PubKey, b64Sig, memo)

	fmtBroadcastStr = strings.Replace(fmtBroadcastStr, " ", "", -1)
	fmtBroadcastStr = strings.Replace(fmtBroadcastStr, "\t", "", -1)
	fmtBroadcastStr = strings.Replace(fmtBroadcastStr, "\n", "", -1)
	fmtBroadcastStr = strings.Replace(fmtBroadcastStr, "\r", "", -1)
	fmt.Printf("formated broadcast string : %v\n", fmtBroadcastStr)

	fmt.Println("=============== Step 5: hex-string encode and broadcast signed transaction data")
	hexBroadcastStr := hex.EncodeToString([]byte(fmtBroadcastStr))
	reqData := map[string]string{"tx": hexBroadcastStr}
	reqUrl := fmt.Sprintf("http://%s/hs/broadcast", gNodeIpPort)
	contentType := "application/json"
	rsp, err := HttpPost(reqUrl, reqData, contentType)
	if err != nil {
		return err
	}

	fmt.Println("=============  Step 6: check response ,  ")
	fmt.Println(rsp)

	return nil
}

func main() {

	amount := HtdfToSatoshi(0.12345678)
	tx_memo := "htdf_demo_in_golang"
	dstAddress := "htdf1jrh6kxrcr0fd8gfgdwna8yyr9tkt99ggmz9ja2"

	err := Transfer(amount, dstAddress, tx_memo)
	if err != nil {
		fmt.Println(err)
	} else {
		fmt.Println(" every thing is ok. ")
	}

	//then, get transaction details on blockchain  as below ( testchain )
	// curl -X GET  http://htdf2020-test01.orientwalt.cn:1317/txs/03FAD07F3B8B521FD4423BC8BC99A8E7F5CA76A8FD8253D8BD7FD2BA6261AA54

}

File: demos/nodejs_demos/htdf_transfer.js
/**
 * author: yqq
 * date: 2020-09-28
 * descriptions: HTDF transaction demo
 * reference : https://gitee.com/orientwalt/apidoc_2020/blob/master/demo/python/htdf_transfer.py
 */

const secp256k1 = require('secp256k1')
const sha256 = require('js-sha256')
const {Base64} = require('js-base64');
const request = require('request')
const utils = require('util')
const bech32 = require('bech32');
const { time } = require('console');
const { chmod } = require('fs');

const gChainId = "testchain"    // mainchain or testchain
const gFromAddress = "htdf1xwpsq6yqx0zy6grygy7s395e2646wggufqndml"
const gPrivateKey = "279bdcd8dccec91f9e079894da33d6888c0f9ef466c0b200921a1bf1ea7d86e8"
const gNodeIpPort = "htdf2020-test01.orientwalt.cn:1317"
const gGasLimit = 30000
const gGasPrice =  100


// mainchain parameters
//const gChainId = "mainchain"                                     //TODO:  mainchain or testchain
//const gFromAddress = "YOUR_HTDF_ADDRESS"                         //TODO
//const gPrivateKey = "YOUR_HTDF_ADDRESS_PRIVATE_KEY"              //TODO
//const gNodeIpPort = "YOUR_HTDF_FULL_NODE_IP_ADDRESS:1317" 		 //TODO
//const gGasLimit = 30000
//const gGasPrice =  100




const gUnsignedStr = `{
	"account_number": "%d",
	"chain_id": "%s",
	"fee": {
		"gas_price": "%d",
		"gas_wanted": "%d"
	},
	"memo": "%s",
	"msgs": [{
		"Amount": [{
		"amount": "%d",
		"denom": "satoshi"
	}],
		"Data": "",
		"From": "%s",
		"GasPrice": %d,
		"GasWanted": %d,
		"To": "%s"
	}],
	"sequence": "%d"
	}`

const gBroadcastStr = `{
	"type": "auth/StdTx",
		"value":{
		"msg": [{
			"type": "htdfservice/send",
			"value":{
				"From": "%s",
				"To": "%s",
				"Amount": [{
					"denom": "satoshi",
					"amount": "%d"
				}],
				"Data": "",
				"GasPrice": "%d",
				"GasWanted": "%d"
			}
		}],
	"fee": {
		"gas_wanted": "%d",
		"gas_price": "%d"
	},
	"signatures": [{
		"pub_key": {
		"type": "tendermint/PubKeySecp256k1",
		"value": "%s"
	},
	"signature": "%s"
	}],
	"memo": "%s"
	}
	}`



// check address validations
function isValidHtdfAddress(address) {
    if(address.length !== 43 && address.toLocaleLowerCase() === address) {
        return false
    }

    try{
        let {prefix , words} =  bech32.decode(address)
        if(prefix === "htdf" && words.length === 32 ){
            return true
        }
    }catch (e) {
    }
    return false
}



function  Transfer( amount , dstAddress , memo   ) {
    // let memo = "htdf_demo_in_golang"

    console.log("===========> step 0 : check parameters")
    if (amount < 1 || amount > 60_000_000 ) {
        throw new Error("nvalid amount " + amount)
    }

    if (!isValidHtdfAddress(dstAddress) ) {
        throw new Error("invalid dstAddress: " + dstAddress)
    }

    console.log("=========> step 1 : get account details")
    let url = utils.format("http://%s/auth/accounts/%s" , gNodeIpPort , gFromAddress )
    let options = {
        method : 'GET',
        headers: { 'Content-Type': 'application/json' }
    }

    request(url, options , function (error, response, body) {

        if (error || response.statusCode !== 200) {
            throw new Error(response.body)
        }

        let accInfo = JSON.parse( body )
        let accountNumber = Number( accInfo["value"]["account_number"] )
        let sequence = Number(accInfo["value"]["sequence"])
        let balance = Number(accInfo["value"]["coins"][0]["amount"])

        let fmtUnsignedStr = utils.format(gUnsignedStr, accountNumber, gChainId, gGasPrice,
            gGasLimit, memo, amount, gFromAddress, gGasPrice, gGasLimit,  dstAddress, sequence)

        fmtUnsignedStr = fmtUnsignedStr.replace(/ /g, '')
        fmtUnsignedStr = fmtUnsignedStr.replace(/\t/g, '')
        fmtUnsignedStr = fmtUnsignedStr.replace(/\n/g, '')
        fmtUnsignedStr = fmtUnsignedStr.replace(/\r/g, '')

        console.log("formatted unsigned string :", fmtUnsignedStr)

        console.log("===========> step 2 : make signature ")

        let sh256 = new sha256.create()
        let msgHash = sh256.update(fmtUnsignedStr).digest()
        console.log("msgHash: ", Buffer.from(msgHash).toString("hex") )

        let privKey = Buffer.from(gPrivateKey, "hex")
        let sig = secp256k1.ecdsaSign(Buffer.from( msgHash), privKey)
        let b64Sig = Base64.encode(sig.signature)
        console.log("base64 encoded signature : ", b64Sig)


        console.log("============> step 3 : make broadcast string ")
        let pubkey = secp256k1.publicKeyCreate(privKey)
        console.log( Buffer.from( pubkey).toString("hex") )

        let b64PubKey = Base64.encode( Buffer.from(pubkey) )

        let fmtBroadcastStr = utils.format(gBroadcastStr, gFromAddress, dstAddress, amount,
            gGasPrice, gGasLimit, gGasLimit, gGasPrice, b64PubKey, b64Sig , memo)

        fmtBroadcastStr = fmtBroadcastStr.replace(/ /g, '')
        fmtBroadcastStr = fmtBroadcastStr.replace(/\t/g, '')
        fmtBroadcastStr = fmtBroadcastStr.replace(/\n/g, '')
        fmtBroadcastStr = fmtBroadcastStr.replace(/\r/g, '')

        console.log("formatted broadcast string: ", fmtBroadcastStr)



        console.log("=========> step 4: broadcast  ")
        let hexBroadcastStr = Buffer.from(fmtBroadcastStr).toString('hex')
        console.log("hex hexBroadcastStr : ", hexBroadcastStr )

        let reqData = {"tx" : hexBroadcastStr}
        url = utils.format("http://%s/hs/broadcast", gNodeIpPort )
        options = {
            method : 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(reqData)
        }

        request(url,  options, function (error, response, body) {
            if (error || response.statusCode !== 200) {
                throw new Error(response.body)
            }

            let rsp = JSON.parse( body )

            if(rsp["code"] ) {
                throw new Error( rsp["raw_log"]  ) // error msg
            }

            console.log("=================")
            console.log("broadcast success : ", rsp)

            console.log("=================")
            tx_url = "http://htdf2020-test01.orientwalt.cn:1317/txs/"
            tx_url += rsp["txhash"]
            console.log( tx_url )
            
            //then, get transaction details on blockchain  as below ( testchain )
            // curl http://htdf2020-test01.orientwalt.cn:1317/txs/930CCBD108B085F4D40CDFE3BD4FF81E7F4B5DA9EC462325ADF5CFB6A4C4F12F
        })
    })
}

function HtdfToSatoshi(htdfamount) {
    // 1 HTDF  = 10^8 satoshi
    return parseInt(htdfamount * 1_0000_0000)
}


function test() {
    try{
        let dstAddress = "htdf1jrh6kxrcr0fd8gfgdwna8yyr9tkt99ggmz9ja2"
        let amount = HtdfToSatoshi(0.1234567)
        let memo = "htdf_transfer_demo_in_js"

        Transfer( amount, dstAddress, memo)
    }catch (e) {
        console.error(e)
    }
}


test()

